File: ios/IonRemote/Utilities/NewConversationRouting.swift
import Foundation

// MARK: - New Conversation Routing
//
// Pure routing logic for the single "New Conversation" entry point
// (conversation unification #256).
//
// Mirrors `resolveNewConversationAction` in
// desktop/src/renderer/components/new-conversation-routing.ts.
// Extracted into its own module so it can be unit-tested without any
// SwiftUI, view-model, or network dependencies.
//
// State machine (highest to lowest precedence, matching desktop):
//   0. Enterprise-locked: policy present and locked=true -> use mandated
//      baseDirectory + profileId. Empty profileId means plain.
//      The enterprise policy reaches iOS over the wire via the
//      desktop_settings_snapshot event (newConversationPolicy field), is stored on
//      SessionViewModel, and is passed in here as `enterprisePolicy`. This
//      branch is gated on policy presence — when `enterprisePolicy` is nil
//      (no policy projected) the branch is skipped.
//   1. Zero engine profiles -> plain conversation, no picker.
//   2. defaultEngineProfileId is non-empty AND the profile exists
//      -> use that profile directly, no picker.
//      (Empty defaultEngineProfileId = "use plain conversation as default"
//       on desktop; falls through to picker here to match intent.)
//   3. Otherwise -> show the extended picker (plain option + profiles).

/// The resolved action the UI should take when the user taps "New Conversation".
enum NewConversationAction: Equatable {
    /// Open a plain (no-extension) conversation tab directly.
    case plain
    /// Open a conversation tab with the given engine profile id directly.
    case profile(profileId: String)
    /// Show the conversation picker (plain + all profiles).
    case showPicker
    /// Enterprise-locked: use the mandated directory and profile id.
    /// profileId="" means plain conversation.
    case locked(baseDirectory: String, profileId: String)
}

/// Enterprise new-conversation policy. Mirrors `NewConversationDefaultsPolicy` in
/// `desktop/src/shared/types-session.ts`. Projected by the desktop as
/// `newConversationPolicy` in the `desktop_settings_snapshot` event, decoded in
/// `NormalizedEvent+EngineDecoder.swift`, stored on
/// `SessionViewModel.enterpriseNewConversationPolicy`, and consumed in `TabListView`.
struct NewConversationDefaultsPolicy: Equatable {
    var locked: Bool
    var baseDirectory: String
    var profileId: String
}

/// Resolve the next action for "New Conversation" given the current
/// preference state. Pure and side-effect-free.
///
/// - Parameters:
///   - profiles:          The current engine profiles list.
///   - defaultId:         The current `defaultEngineProfileId` (empty = unset).
///   - enterprisePolicy:  The enterprise NewConversationDefaults policy, or nil if none.
func resolveNewConversationAction(
    profiles: [EngineProfile],
    defaultId: String,
    enterprisePolicy: NewConversationDefaultsPolicy? = nil
) -> NewConversationAction {
    // State 0 (highest precedence): enterprise-locked.
    if let policy = enterprisePolicy, policy.locked {
        return .locked(baseDirectory: policy.baseDirectory, profileId: policy.profileId)
    }

    // State 1: no profiles at all -> plain conversation, no picker.
    if profiles.isEmpty { return .plain }

    // State 2: a default is set and the profile still exists -> use it.
    if !defaultId.isEmpty {
        let exists = profiles.contains(where: { $0.id == defaultId })
        if exists { return .profile(profileId: defaultId) }
        // Default was deleted: fall through to picker.
    }

    // State 3: show the extended picker (plain + profiles).
    return .showPicker
}

File: desktop/src/renderer/components/new-conversation-routing.ts
import type { EngineProfile, NewConversationDefaultsPolicy } from '../../shared/types'

/**
 * new-conversation-routing — pure routing logic for the single "New
 * Conversation" entry point (conversation unification #256).
 *
 * Extracted into its own module so it can be unit-tested without any
 * React, DOM, Electron, or store dependencies.
 */

export type NewConversationAction =
  | { kind: 'plain' }
  | { kind: 'profile'; profileId: string }
  | { kind: 'show-picker' }
  | { kind: 'locked'; baseDirectory: string; profileId: string }

/**
 * Resolve the next action for "new conversation" given the current preference
 * state. Pure and side-effect-free.
 *
 * State machine (highest to lowest precedence):
 *   0. Enterprise-locked (highest): NewConversationDefaults.locked is true -> skip both
 *      pickers. Open directly with the mandated baseDirectory and profileId.
 *      Empty profileId means plain conversation.
 *   1. Zero engine profiles -> open plain conversation directly. No picker.
 *   2. defaultEngineProfileId is non-empty AND the profile still exists
 *      -> open with that profile directly. No picker.
 *   3. Otherwise -> show the extended picker (plain option + profiles).
 *
 * @param profiles         The current engineProfiles list.
 * @param defaultId        The current defaultEngineProfileId preference ('' = unset).
 * @param enterprisePolicy The enterprise NewConversationDefaults policy, or null if none.
 */
export function resolveNewConversationAction(
  profiles: EngineProfile[],
  defaultId: string,
  enterprisePolicy?: NewConversationDefaultsPolicy | null,
): NewConversationAction {
  // State 0 (highest precedence): enterprise-locked. Skip both pickers entirely.
  if (enterprisePolicy?.locked) {
    return {
      kind: 'locked',
      baseDirectory: enterprisePolicy.baseDirectory ?? '',
      profileId: enterprisePolicy.engineProfileId ?? '',
    }
  }

  // State 1: no profiles at all -> plain conversation, no picker.
  if (profiles.length === 0) return { kind: 'plain' }

  // State 2: a default is set and the profile still exists -> use it directly.
  if (defaultId) {
    const exists = profiles.some((p) => p.id === defaultId)
    if (exists) return { kind: 'profile', profileId: defaultId }
    // Default was deleted: fall through to picker.
  }

  // State 3: show the extended picker.
  return { kind: 'show-picker' }
}

/**
 * Execute the resolved routing action for a specific directory. Handles the
 * `locked`, `plain`, and `profile` branches by calling the appropriate store
 * action. Returns `'show-picker'` when the caller needs to open the
 * `NewConversationPicker` UI (the caller handles that).
 *
 * This eliminates duplication across TabStrip, useKeyboardShortcuts, context
 * menus, and the pill creation path. Every call site that creates a new
 * conversation tab MUST route through this helper (or through
 * `resolveNewConversationAction` directly) so the enterprise lock cannot be
 * bypassed.
 *
 * @param dir               Working directory to use (fallback from preferences
 *                          when empty/undefined).
 * @param action            The pre-resolved action from
 *                          `resolveNewConversationAction`.
 * @param createTabInDir    Store action: `sessionStore.createTabInDirectory`.
 * @param createConvTab     Store action: `sessionStore.createConversationTab`.
 * @param shouldUseWorktree Whether to open a worktree tab.
 * @returns 'show-picker' when the caller must present the picker UI,
 *          'done' when the tab was created.
 */
export function executeNewConversationAction(
  dir: string,
  action: NewConversationAction,
  createTabInDir: (dir: string, worktree?: boolean) => void,
  createConvTab: (dir: string, opts?: { profileId?: string }) => void,
  shouldUseWorktree: boolean = false,
): 'show-picker' | 'done' {
  switch (action.kind) {
    case 'locked': {
      // Enterprise-locked: always use the mandated dir (fall back to provided
      // dir if the mandated one is empty, matching desktop's new-tab behavior).
      const lockedDir = action.baseDirectory || dir
      if (action.profileId) {
        createConvTab(lockedDir, { profileId: action.profileId })
      } else {
        createTabInDir(lockedDir, shouldUseWorktree)
      }
      return 'done'
    }
    case 'plain':
      createTabInDir(dir, shouldUseWorktree)
      return 'done'
    case 'profile':
      createConvTab(dir, { profileId: action.profileId })
      return 'done'
    case 'show-picker':
      return 'show-picker'
  }
}

/**
 * newTabInDirectory — the single, lock-safe "new tab in this directory" entry
 * point shared by every per-tab context menu ("New tab in dir" on the tab
 * context menu, the group-pill context menu, and the group-picker dropdown's
 * per-tab menu).
 *
 * This collapses what used to be three duplicated inline handlers — two of
 * which called `createTabInDirectory` directly and so bypassed the enterprise
 * lock entirely. Routing every context-menu "new tab in dir" path through this
 * one function means the lock cannot be bypassed: there is exactly one path,
 * and it always resolves the action (including the highest-precedence `locked`
 * branch) before creating anything.
 *
 * It reads the routing inputs (profiles, default profile, enterprise policy)
 * from the preferences store and dispatches to the session store, so callers
 * only supply the target directory and worktree flag.
 *
 * @param dir   The tab's working directory to open the new tab in.
 * @param deps  The store accessors + worktree flag (injected so this is unit-
 *              testable without React/Electron).
 * @returns 'show-picker' when the resolver wants the picker UI (the caller is
 *          a context menu with no picker, so this is effectively 'no-op' there),
 *          'done' when a tab was created.
 */
export function newTabInDirectory(
  dir: string,
  deps: {
    profiles: EngineProfile[]
    defaultProfileId: string
    enterprisePolicy?: NewConversationDefaultsPolicy | null
    createTabInDir: (dir: string, worktree?: boolean) => void
    createConvTab: (dir: string, opts?: { profileId?: string }) => void
    shouldUseWorktree?: boolean
  },
): 'show-picker' | 'done' {
  const action = resolveNewConversationAction(
    deps.profiles,
    deps.defaultProfileId,
    deps.enterprisePolicy,
  )
  return executeNewConversationAction(
    dir,
    action,
    deps.createTabInDir,
    deps.createConvTab,
    deps.shouldUseWorktree ?? false,
  )
}
